src/ring_theory/subsemiring.lean

```diff
@@ -34,6 +35,61 @@ add_decl_doc subsemiring.to_submonoid
 /-- Reinterpret a `subsemiring` as an `add_submonoid`. -/
 add_decl_doc subsemiring.to_add_submonoid
 
+namespace submonoid
+
+variables {a b : R}
+
+/-- The product of an element of the additive closure of a multiplicative submonoid `M`
+and an element of `M` is contained in the additive closure of `M`. -/
+lemma mul_right_mem_add_closure
+  (ha : a ∈ add_submonoid.closure (M : set R)) (hb : b ∈ M) :
+  a * b ∈ add_submonoid.closure (M : set R) :=
+begin
+  revert b,
+  refine add_submonoid.closure_induction ha _ _ _; clear ha a,
+  { exact λ r hr b hb, add_submonoid.mem_closure.mpr (λ y hy, hy (M.mul_mem hr hb)) },
+  { exact λ b hb, by simp only [zero_mul, (add_submonoid.closure (M : set R)).zero_mem] },
+  { simp_rw add_mul,
+    exact λ r s hr hs b hb, (add_submonoid.closure (M : set R)).add_mem (hr hb) (hs hb) }
+end
+
+/-- The product of an element of `M` and an element of the additive closure of a multiplicative
+submonoid `M` is contained in the additive closure of `M`. -/
+lemma mul_left_mem_add_closure (ha : a ∈ M) (hb : b ∈ add_submonoid.closure (M : set R)) :
+  a * b ∈ add_submonoid.closure (M : set R) :=
+begin
+  revert a,
+  refine add_submonoid.closure_induction hb _ _ _; clear hb b,
+  { exact λ r hr b hb, add_submonoid.mem_closure.mpr (λ y hy, hy (M.mul_mem hb hr)) },
+  { exact λ b hb, by simp only [mul_zero, (add_submonoid.closure (M : set R)).zero_mem] },
+  { simp_rw mul_add,
+    exact λ r s hr hs b hb, (add_submonoid.closure (M : set R)).add_mem (hr hb) (hs hb) }
+end
+
+variable {M}
+
+/-- The product of two elements of the additive closure of a submonoid `M` is an element of the
+additive closure of `M`. -/
+lemma mul_mem_add_closure
+  (ha : a ∈ add_submonoid.closure (M : set R)) (hb : b ∈ add_submonoid.closure (M : set R)) :
+  a * b ∈ add_submonoid.closure (M : set R) :=
+begin
+  revert a,
+  refine add_submonoid.closure_induction hb _ _ _; clear hb b,
+  { exact λ r hr b hb, M.mul_right_mem_add_closure hb hr },
+  { exact λ b hb, by simp only [mul_zero, (add_submonoid.closure (M : set R)).zero_mem] },
+  { simp_rw mul_add,
+    exact λ r s hr hs b hb, (add_submonoid.closure (M : set R)).add_mem (hr hb) (hs hb) }
+end
```

Comment:
If you swap the order of these proofs, can't you prove `mul_left_mem_add_closure` in terms of `mul_mem_add_closure`?